### ISSUE TITLE
Docker Containers offers download stream if html template is accessed with nano while container is running

### ISSUE THREAD

**Issue by mfld-pub** [mpolden/echoip#138]
I was running the previous docker image that was from early 2020 and  would provide it a single index.html file to serve with the -t parameter.

With that 2020 version it was possible to edit the .html file with the service running and changes would be reflected real-time on next visit / browser refresh. 

I now updated to the latest image, the one from February 2021.
 
        "Id": "sha256:b8a3c104043e1f9bf0c2c94577ca6323642b90aee74cf81a38eeec14ff3f6206",
        "RepoTags": [
            "mpolden/echoip:latest"
        ],
        "RepoDigests": [
            "mpolden/echoip@sha256:b9231d6a919cf3aeb11b9cbccfb20b274afc1cb0ac9797f2c65fca1f421449f2"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2021-02-24T20:00:23.546693989Z",

With this update, the -t parameter seems to point to a directory in which to place .html rather than a discrete file. Thus I created a new directory /static and placed my only index.html in it. 

Now when the container is running it is picking up my index.html and serving just fine but if I open this .html file in nano while the service is running subsequent visitors to the site will be offered a binary stream to download. Observed in Chrome, Firefox and Edge. 

Once I close nano subsequent request are served the correct content again. Bizarre. It does not happen if I open the .html file in vi and press "insert". Only with Nano.

Developer console shows: `Resource interpreted as Document but transferred with MIME type application/octet-stream`

`docker logs echoip `shows nothing out of the ordinary.

### Desired behavior:

Accessing a .html template from the designated template directory should not impact service.

### Actual behavior:

Merely opening the .html in nano without making any changes causes the service to offer a binary stream to download for as long as the .html is open in nano. It does not happen with vi. 

### Steps to reproduce:

1) Ubuntu 20.04, Docker CE  Version:  20.10.6, NGINX reverse proxy, Cloudflare.

2) docker pull mpolden/echoip:latest

3) Place an index.html file into /var/lib/GeoIP/static/

4) docker run --rm --name=echoip -d -v /var/lib/GeoIP:/geo --network host mpolden/echoip -f /geo/GeoLite2-Country.mmdb -c /geo/GeoLite2-City.mmdb -a /geo/GeoLite2-ASN.mmdb -p -r -t /geo/static/ -H X-Forwarded-For

5) Content should now be served as expected

6) ~~Using GNU nano, version 4.8 open the index.html file `nano /var/lib/GeoIP/static/index.html`~~
    Place any file that is NOT html and begins with [0-9,a-h] into /var/lib/GeoIP/static/

7) Control / F5 the browser. Content will not be served. It will offer a download of the file added above.

8) ~~Having made no changes, close nano.~~
   Remove the file added above or rename it so it begins with [j-z].

9) Refresh browser once more. Content is being served again.

I notice curl requests are still being handled correctly while this is going on.

**Comment by maximbaz** [comment]
Fun report! 😄 

It sounds like it's just `nano` that is doing something to the file when it opens it, especially as you say it doesn't reproduce with `vi`. Interestingly I cant reproduce this issue! The only difference is that my nano is v5.7, so it could be that this is something they fixed later.

Out of curiosity, could you try to run maybe `stat static/index.html` and `file -bi static/index.html` when nano is open and closed and see if there is any difference?

**Comment by mfld-pub** [comment]
> Fun report! 😄

Crazy, right ?! 

I just replicated it on a fresh instance. It is definitely a thing. I wonder what change in the latest docker image causes this.

I retained snapshots of new and old, spun them up separately so I can test for this ticket.

With latest image running and in production, nano not touching the .html:

```
file -bi  /var/lib/GeoIP/static/index.html
text/html; charset=us-ascii

stat /var/lib/GeoIP/static/index.html
  File: /var/lib/GeoIP/static/index.html
  Size: 3218            Blocks: 8          IO Block: 4096   regular file
Device: 801h/2049d      Inode: 133991      Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-18 15:02:19.359060046 +0000
Modify: 2021-05-18 15:02:12.790821186 +0000
Change: 2021-05-18 15:02:12.790821186 +0000
 Birth: -

```

Now doing `nano /var/lib/GeoIP/static/index.html`

Site is broken.

```
file -bi  /var/lib/GeoIP/static/index.html
text/html; charset=us-ascii

stat /var/lib/GeoIP/static/index.html
  File: /var/lib/GeoIP/static/index.html
  Size: 3218            Blocks: 8          IO Block: 4096   regular file
Device: 801h/2049d      Inode: 133991      Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-18 15:02:19.359060046 +0000
Modify: 2021-05-18 15:02:12.790821186 +0000
Change: 2021-05-18 15:02:12.790821186 +0000
 Birth: -
```

I notice that while the nano currently shipping with Ubuntu 20.04 has a file open, it creates a hidden temporary file. 

.index.html.swp

Perhaps the new version of echoip tries to parse everything in the template directory, regardless of extension or if it is hidden with a dot prefix or not.

**Comment by mfld-pub** [comment]
Update: Yes! Mystery solved. 

I can break things by simply putting any non HTML file into the designate templates directory.

touch /var/lib/GeoIP/static/.BREAKME.txt 

results in service displaying a blank page.

Removing this empty file restores service.

`cp /usr/bin/cal /var/lib/GeoIP/static/`

Results in Internal Server Error.

Bottom Line: Current echoip image seems to process and try to serve up anything placed into the templates directory that you specify with -t. It does not sanitize for file type, file name or content.

Interesting. `cp /etc/passwd /var/lib/GeoIP/static/` does not harm anything. Perhaps it parses the contents of templates directory and will process in alphabetical order ?

Yep.

`cp /etc/passwd /var/lib/GeoIP/static/H-comes-before-i`

Now displays my /etc/passwd in plain text.

**Comment by maximbaz** [comment]
Nice analysis! Yeah I believe when serving index.html at least, `echoip` should serve all resources with correct Content-Type.

**Comment by mfld-pub** [comment]
> Nice analysis! Yeah I believe when serving index.html at least, `echoip` should serve all resources with correct Content-Type.

Can you replicate it now ? Basically start container with -t /some/path, drop an index.html in there. It will serve it. Now drop some binary file that begins with [0-h] into /some/path. Refresh browser. It will try serve all files in /some/path in alphabetical order. Anything not HTML will be processed anyway and sent as-is.

**Comment by maximbaz** [comment]
Yeah. Hm it's not nice that we try to parse random files (especially if it causes problems), but it's just how the `html/template` library works (or how it's used here) - we point it to a directory and tell it to parse each and every file in that directory and serve it:

https://github.com/mpolden/echoip/blob/92a434d7eafdd353415c7e819e32dd06cf878a68/http/http.go#L326

I thought of changing the glob e.g. to only include `.html` files, but that won't work, if you added a `.css` file you _want_ it to be processed - and in general we do want to allow people to create and serve whichever files they want. And what's even worse, if we include a glob for an extension that the folder will not have, this will crash the library 🤦‍♂️ 

I don't have any obvious ideas how to fix this, besides documenting "make sure the folder does not contain any temp files or any files that you dont want to be served" 😕
